components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineView.kt

```diff
@@ -471,9 +516,51 @@ class SystemEngineView @JvmOverloads constructor(
         }
     }
 
+    private fun resetJSAlertAbuseState() {
+        alertJSCount = 0
+        userWantToShowMoreDialogs = true
+    }
+
+    internal fun updateJSDialogAbusedState() {
+        if (!areDialogsBeingAbusedByTime()) {
+            ++alertJSCount
+            lastTimeWhenAJSDialogWasShown = Date()
+        } else {
+            alertJSCount = 0
+        }
+    }
+
+    internal fun areDialogsBeingAbused(): Boolean {
+        return areDialogsBeingAbusedByTime() || areDialogsBeingAbusedByAmount()
+    }
+
+    @Suppress("MagicNumber")
+    internal fun areDialogsBeingAbusedByTime(): Boolean {
+        return if (alertJSCount == 0) {
+            false
+        } else {
+            val now = Date()
+            val diffInSeconds =
+                (now.time - lastTimeWhenAJSDialogWasShown.time) / 1000 // 1 second equal to 1000 milliseconds
+
+            diffInSeconds > MAX_SUCCESSIVE_DIALOG_SECONDS_LIMIT
+        }
+    }
+
+    internal fun areDialogsBeingAbusedByAmount(): Boolean {
+        return alertJSCount > MAX_SUCCESSIVE_DIALOG_COUNT
+    }
+
     override fun canScrollVerticallyDown() = currentWebView.canScrollVertically(1)
 
     companion object {
+
+        // Maximum number of successive dialogs before we prompt users to disable dialogs.
+        internal const val MAX_SUCCESSIVE_DIALOG_COUNT: Int = 2
+
+        // Maximum Amount of seconds allowed between js alert before enabling the stop dialog.
```

Comment:
// Minimum time required between dialogs in seconds ... would make this clearer see comment above about the logic being inverted.